### PR TITLE
Add API for sender UUIDs and fix CB not using sender UUID in chat

### PR DESCRIPTION
**Pull request by Proximyst**
This PR does two things (both to achieve the same goal, though):

- Add a new API to `Player` to set the sender UUID of a message that is sent (as added in 20w21a);
- Fix CraftBukkit not using the sender UUID of chat when handling chat messages. Kudos to @egg82 for pointing this out.

### CODE REVIEW

File: Spigot-API-Patches/0231-Add-senderUuid-to-sendMessage.patch [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+     @NotNull
+     <T> T getClientOption(@NotNull ClientOption<T> option);
++
++    /**
++     * Sends this sender a message raw
++     *
++     * @param message Message to be displayed
++     * @param senderUuid The sender of this message; null for a system message
++     */
++    void sendRawMessage(@NotNull String message, @Nullable UUID senderUuid);
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Add-senderUuid-to-sendMessage.patch]: Cat: The params should be `UUID, MessageType`.

**Comment by kennytv** [comment]
The sender uuid doesn't really have any purpose atm (other than 0,0 for unhidable 'console' or a player's own uuid) 👀

**Comment by Proximyst** [comment]
At the moment, no. It may in the future, though; perhaps blocking people can become a per-user feature in their clients? Who knows!

**Review by aikar — COMMENTED** [review]
Overall looks good but I feel like UUID should be after message not before, so that it is more of an overloaded version than it is a completely different signature.

Thoughts?

**Comment by Proximyst** [comment]
Cat felt the other way around, hence my change on it. Having the UUID first also lets you use varargs for `BaseComponent` messages, as some may want.

File: Spigot-API-Patches/0231-Add-senderUuid-to-sendMessage.patch [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
++    /**
++     * Sends an array of components as a single message to the player
++     *
++     * <p>If this player does not support sending full components then
++     * the components will be sent as legacy text.</p>
++     *
++     * @param components the components to send
++     * @param senderUuid The sender of this message; null for a system message
++     */
++    default void sendMessage(@Nullable UUID senderUuid, @NotNull net.md_5.bungee.api.chat.BaseComponent[] components) {
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Add-senderUuid-to-sendMessage.patch:70]:
```suggestion
+    default void sendMessage(@Nullable UUID senderUuid, @NotNull net.md_5.bungee.api.chat.BaseComponent... components) {
```

**Comment by aikar** [comment]
@KennyTV it is used by the client if the client requests to disable chat.

So a parent can tell the client for their child to not show chat, but this bug blocks that ability.

With this change, at least a VANILLA chat experience will function as vanilla.

Then will just have to get chat plugins to honor it too.

Plus, client mods could be created to do fancy stuff like hovering over chat message and pop up details about the player who sent it.

Mods will be less encouraged to be created if the system doesnt work at all :P

**Comment by aikar** [comment]
I guess varargs is a reasonable reason.

**Comment by egg82** [comment]
@KennyTV the sender does have a use for current Notchian clients >= 20w21a.

**Comment by kennytv** [comment]
Yes, but as I said it's only checked against the console or the player's own uuid and not customizable outside of that... buuut I agree with proxi and aikar in that it should work as intended in case it has any use outside of that one client flag in the future / for mods

**Review by kashike — DISMISSED** [review]

File: Spigot-API-Patches/0231-Add-senderUuid-to-sendMessage.patch [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
++    void sendMessage(@Nullable UUID senderUuid, @NotNull String[] messages);
++
++    /**
++     * Sends the component to the player
++     *
++     * <p>If this player does not support sending full components then
++     * the component will be sent as legacy text.</p>
++     *
++     * @param component the component to send
++     * @param senderUuid The sender of this message; null for a system message
```
Comment (mbax) [inline Spigot-API-Patches/0231-Add-senderUuid-to-sendMessage.patch:55]: Super nitpicky but there's inconsistent capitalization after parameter names throughout this and here is one example <3

**Comment by Proximyst** [comment]
PR marked as draft while this is attempted upstream.

**Comment by Proximyst** [comment]
Merged upstream.

**Pull request closed without merging**